File: api_anita.py
"""
Cliente Python del bridge Anita (equivalente a App\\ApiAnita de anitaERP).

Configuración por variables de entorno (mismas que config/anita.php):
  ANITA_IP, LOCAL_IP, ANITA_BRIDGE_TYPE, ANITA_API_SCRIPT,
  ANITA_BDD, ANITA_BDD_PATH, IFX_SERVER, IFX_SERVER_LOCAL,
  ANITA_PUERTO_SSH

Uso:
    from api_anita import ApiAnita

    api = ApiAnita()
    resp = api.api_call({
        "acc": "list",
        "tabla": "clientes",
        "campos": "id, nombre",
        "whereArmado": "WHERE id = 1",
    })
"""

from __future__ import annotations

import json
import os
import random
import re
import subprocess
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Mapping, MutableMapping, Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

try:
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:
    pass


def _env(key: str, default: str = "") -> str:
    return (os.environ.get(key) or default).strip()


@dataclass
class AnitaConfig:
    """Valores equivalentes a config/anita.php."""

    ip: str = field(default_factory=lambda: _env("ANITA_IP"))
    local_ip: str = field(default_factory=lambda: _env("LOCAL_IP"))
    bridge_type: str = field(
        default_factory=lambda: _env("ANITA_BRIDGE_TYPE", "HTTP") or "HTTP"
    )
    api_script: str = field(
        default_factory=lambda: _env("ANITA_API_SCRIPT", "apiERP.php") or "apiERP.php"
    )
    bdd: str = field(default_factory=lambda: _env("ANITA_BDD", "ventas") or "ventas")
    bdd_path: str = field(default_factory=lambda: _env("ANITA_BDD_PATH"))
    ifx_server: str = field(default_factory=lambda: _env("IFX_SERVER"))
    ifx_server_local: str = field(default_factory=lambda: _env("IFX_SERVER_LOCAL"))
    puerto_ssh: Optional[str] = field(default_factory=lambda: _env("ANITA_PUERTO_SSH") or None)
    servidores: dict[str, str] = field(default_factory=dict)
    ifx_servers: dict[str, str] = field(default_factory=dict)
    logs_dir: str = field(default_factory=lambda: os.path.join(os.getcwd(), "logs"))
    ssh_user: str = "sergio"

    def __post_init__(self) -> None:
        if not self.servidores:
            self.servidores = {
                "ANITA_IP": self.ip,
                "LOCAL_IP": self.local_ip,
            }
        if not self.ifx_servers:
            self.ifx_servers = {
                "IFX_SERVER": self.ifx_server,
                "IFX_SERVER_LOCAL": self.ifx_server_local,
            }


class ApiAnita:
    """Bridge HTTP (por defecto) o legacy SSH+SCP hacia Informix vía apiERP.php."""

    _HOST_PATTERN = re.compile(r"^[\w.\-]+(:\d+)?$", re.ASCII)

    def __init__(
        self,
        config: Optional[AnitaConfig] = None,
        *,
        servidor: Optional[str] = None,
    ) -> None:
        self._config = config or AnitaConfig()
        micro = datetime.now().strftime("%Y%m%d%H%M%S%f")[:-3]
        self.fecha = f"{micro}_{random.randint(0, 9999)}"
        self.servidor_anita = servidor if servidor is not None else self._config.ip

    @staticmethod
    def resolver_host(
        clave_servidor: Optional[str] = None,
        config: Optional[AnitaConfig] = None,
    ) -> str:
        cfg = config or AnitaConfig()
        if clave_servidor is not None and clave_servidor.strip():
            clave = clave_servidor.strip().upper()
            if clave in cfg.servidores and cfg.servidores[clave].strip():
                return cfg.servidores[clave].strip()
            if ApiAnita._HOST_PATTERN.match(clave_servidor.strip()):
                return clave_servidor.strip()
        return cfg.ip.strip()

    @staticmethod
    def url_bridge(host: Optional[str] = None, config: Optional[AnitaConfig] = None) -> str:
        cfg = config or AnitaConfig()
        resolved = (host or ApiAnita.resolver_host(config=cfg)).strip()
        if not resolved:
            raise RuntimeError(
                "ANITA_IP no está configurado. Defínalo en .env o en AnitaConfig.ip."
            )

        script = cfg.api_script.lstrip("/")
        if re.match(r"^https?://", resolved, re.IGNORECASE):
            base = resolved.rstrip("/")
            if re.search(r"\.php(\?|$)", base, re.IGNORECASE):
                return base
            return f"{base}/{script}"
        return f"http://{resolved}/{script}"

    @staticmethod
    def _resolver_ifx_server(
        clave_ifx: Optional[str] = None,
        config: Optional[AnitaConfig] = None,
    ) -> str:
        cfg = config or AnitaConfig()
        if clave_ifx is not None and clave_ifx.strip():
            clave = clave_ifx.strip().upper()
            if clave in cfg.ifx_servers and cfg.ifx_servers[clave].strip():
                return cfg.ifx_servers[clave].strip()
        return cfg.ifx_server.strip()

    def api_call_http(self, data: Mapping[str, Any]) -> str:
        payload: dict[str, Any] = dict(data)
        acc = str(payload.get("acc", ""))

        if "servidor" in payload:
            self.servidor_anita = self.resolver_host(
                str(payload["servidor"]), self._config
            )

        if "ifx_server" in payload:
            payload["IFX_SERVER"] = self._resolver_ifx_server(
                str(payload["ifx_server"]), self._config
            )
        else:
            payload["IFX_SERVER"] = self._resolver_ifx_server(config=self._config)

        bdd = self._config.bdd
        bdd_path = self._config.bdd_path.rstrip("/")
        payload["DB_NAME"] = bdd
        payload["IFX_DB_PATH"] = f"{bdd_path}/{bdd}" if bdd_path else bdd
        payload["sistema"] = bdd
        if bdd_path:
            payload["path_sistema"] = bdd_path

        try:
            url = self.url_bridge(self.servidor_anita, self._config)
        except RuntimeError as exc:
            return json.dumps({"Error": f"Bridge HTTP Anita: {exc}"})

        body = json.dumps(payload).encode("utf-8")
        request = Request(
            url,
            data=body,
            method="POST",
            headers={
                "Accept": "application/json",
                "Content-Type": "application/json",
            },
        )

        try:
            with urlopen(request, timeout=120) as response:
                raw = response.read()
        except HTTPError as exc:
            try:
                detail = exc.read().decode("utf-8", errors="replace")
            except OSError:
                detail = str(exc)
            return json.dumps({"Error": f"Bridge HTTP Anita: {detail or exc}"})
        except URLError as exc:
            return json.dumps({"Error": f"Bridge HTTP Anita: {exc.reason}"})

        text = raw.decode("utf-8", errors="replace")
        trim = text.strip()
        if not trim:
            if acc == "list":
                return "[]"
            if acc in ("insert", "update", "delete"):
                return "[]"
            return json.dumps({"Error": "Bridge HTTP Anita: respuesta vacía"})
        return text

    @staticmethod
    def _limpiar_respuesta_bridge_escritura(respuesta: str) -> str:
        lineas = []
        for linea in respuesta.splitlines():
            if re.match(r"^\s*(?:Warning|Notice|Deprecated|Fatal error)\b", linea, re.I):
                continue
            lineas.append(linea)
        return "\n".join(lineas).strip()

    @staticmethod
    def _respuesta_bridge_escritura_exitosa(respuesta: str) -> bool:
        limpia = ApiAnita._limpiar_respuesta_bridge_escritura(respuesta)
        return bool(
            limpia
            and re.search(r"\d+\s+row\(s\)\s+(?:inserted|updated|deleted)\b", limpia, re.I)
        )

    @staticmethod
    def extraer_mensaje_error(respuesta: Optional[str]) -> Optional[str]:
        if respuesta is None:
            return "Sin respuesta del bridge Anita"

        trim = respuesta.strip()
        if not trim:
            return "Respuesta vacía del bridge Anita"
        if trim in ("[]", "{}"):
            return None

        try:
            decoded = json.loads(trim)
        except json.JSONDecodeError:
            limpia = ApiAnita._limpiar_respuesta_bridge_escritura(trim)
            if ApiAnita._respuesta_bridge_escritura_exitosa(trim):
                return None
            if limpia.lower() == "error":
                return "Error en ejecución SQL Informix (revise el archivo .ret en el servidor Anita)"
            lower = trim.lower()
            if "<b>warning</b>" in lower or "<b>fatal error</b>" in lower:
                return re.sub(r"<[^>]+>", "", trim)
            if limpia == "" and re.search(r"\b(?:warning|notice|fatal error)\b", trim, re.I):
                return "Advertencia PHP en bridge Anita (actualice apiERP.php en el servidor)"
            if "error" in limpia.lower():
                return limpia
            return None

        if isinstance(decoded, dict):
            err = decoded.get("Error") or decoded.get("error")
            if err is not None and str(err).strip():
                return str(err)
        return None

    @staticmethod
    def decodificar_lista_filas(respuesta: Optional[str]) -> list[dict[str, Any]]:
        if respuesta is None:
            return []

        trim = respuesta.strip()
        if not trim or trim == "[]":
            return []

        try:
            decoded = json.loads(trim)
        except json.JSONDecodeError:
            return []

        if isinstance(decoded, list):
            return [row for row in decoded if isinstance(row, dict)]
        if isinstance(decoded, dict):
            if decoded.get("Error") or decoded.get("error"):
                return []
            return [decoded]
        return []

    @staticmethod
    def primera_fila_lista(respuesta: Optional[str]) -> Optional[dict[str, Any]]:
        filas = ApiAnita.decodificar_lista_filas(respuesta)
        return filas[0] if filas else None

    def _usa_bridge_http(self) -> bool:
        return self._config.bridge_type.strip().upper() == "HTTP"

    def api_call(self, data: Mapping[str, Any]) -> str:
        if self._usa_bridge_http():
            return self.api_call_http(data)
        return self._api_call_ssh_legacy(dict(data))

    def _api_call_ssh_legacy(self, data: MutableMapping[str, Any]) -> str:
        puerto = self._config.puerto_ssh
        port_ssh = f"-p {puerto}" if puerto else ""
        port_scp = f"-P {puerto}" if puerto else ""
        host = self.servidor_anita
        user = self._config.ssh_user

        sql = self.armar_sql(data)
        nom_arch = f"{self.fecha}.sql"
        logs_dir = self._config.logs_dir
        os.makedirs(logs_dir, exist_ok=True)
        path_arch = os.path.join(logs_dir, nom_arch)

        with open(path_arch, "w", encoding="utf-8") as fh:
            fh.write(sql)

        remote_tmp = f"/home/{user}/tmp/{nom_arch}"
        subprocess.run(
            ["scp", *([port_scp] if port_scp else []), path_arch, f"{user}@{host}:{remote_tmp}"],
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        subprocess.run(
            [
                "ssh",
                *([port_ssh] if port_ssh else []),
                f"{user}@{host}",
                (
                    f"cd /usr2/www/htdocs; ./apiERP.php {self._config.bdd} "
                    f"{remote_tmp} {self.fecha} > /dev/null"
                ),
            ],
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        subprocess.run(
            [
                "ssh",
                *([port_ssh] if port_ssh else []),
                f"{user}@{host}",
                f"rm {remote_tmp} > /dev/null",
            ],
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )

        acc = data.get("acc")
        if acc in ("list", "customSql"):
            csv_path = os.path.join(logs_dir, f"{self.fecha}.csv")
            bdd_path = self._config.bdd_path.rstrip("/")
            bdd = self._config.bdd
            remote_csv = f"{bdd_path}/{bdd}/{self.fecha}.csv"
            subprocess.run(
                [
                    "scp",
                    *([port_scp] if port_scp else []),
                    f"{user}@{host}:{remote_csv}",
                    csv_path,
                ],
                check=False,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            subprocess.run(
                [
                    "ssh",
                    *([port_ssh] if port_ssh else []),
                    f"{user}@{host}",
                    f"cd {bdd_path}/{bdd}; rm {self.fecha}.csv > /dev/null",
                ],
                check=False,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )

            if not os.path.isfile(csv_path):
                sql_file = os.path.join(logs_dir, f"{self.fecha}.sql")
                if os.path.isfile(sql_file):
                    os.remove(sql_file)
                return json.dumps([])

            campos = str(data.get("campos", ""))
            campos_arr = [c.strip() for c in campos.split(",") if c.strip()]
            data_arr: list[dict[str, str]] = []

            with open(csv_path, encoding="utf-8", errors="replace") as archivo:
                for linea in archivo:
                    linea_arr = linea.rstrip("\n").split("|")
                    registro: dict[str, str] = {}
                    for key, value in enumerate(campos_arr):
                        nombre = value
                        for sep in (" as ", " AS "):
                            partes = value.split(sep)
                            if len(partes) == 2:
                                nombre = partes[1].strip()
                                break
                        if key < len(linea_arr):
                            registro[nombre.strip()] = linea_arr[key]
                    data_arr.append(registro)

            for path in (csv_path, path_arch):
                if os.path.isfile(path):
                    os.remove(path)
            return json.dumps(data_arr)

        return json.dumps([])

    def armar_sql(self, data: Mapping[str, Any]) -> str:
        where = str(data.get("where", ""))
        campos = str(data.get("campos", ""))
        tabla = str(data.get("tabla", ""))
        where_armado = str(data.get("whereArmado", ""))
        order_by = (
            f" ORDER BY {data['orderBy']}" if data.get("orderBy") else ""
        )
        group_by = (
            f" GROUP BY {data['groupBy']}" if data.get("groupBy") else ""
        )
        valores = str(data.get("valores", ""))
        acc = data.get("acc")

        if acc == "list":
            sql = (
                f"UNLOAD TO '{self.fecha}.csv' DELIMITER '|' SELECT {campos} "
                f"FROM {tabla} {where_armado} {group_by} {order_by}"
            )
        elif acc == "insert":
            sql = f"INSERT INTO {tabla} ({campos}) VALUES ({valores})"
        elif acc == "update":
            sql = f"UPDATE {tabla} SET {valores} {where_armado}"
        elif acc == "delete":
            sql = f"DELETE FROM {tabla} {where_armado}"
        elif acc == "customSql":
            sql = (
                f"UNLOAD TO '{self.fecha}.csv' DELIMITER '|' "
                f"{data.get('customSql', '')}"
            )
        else:
            sql = ""

        return re.sub(r"\s\s+", " ", sql.strip())

    def obtener_siguiente_numerador(
        self, tabla: str, campo_id: str = "id"
    ) -> int:
        payload = {
            "acc": "list",
            "campos": f"MAX({campo_id}) AS id",
            "tabla": tabla,
        }
        fila = self.primera_fila_lista(self.api_call(payload))
        if fila is not None:
            valor = fila.get("id")
            if valor is not None and str(valor).strip() != "":
                return int(valor) + 1
        return 1


# Alias compatible con nombre PHP
ApiAnita.resolverHost = staticmethod(ApiAnita.resolver_host)  # type: ignore[method-assign]
ApiAnita.urlBridge = staticmethod(ApiAnita.url_bridge)  # type: ignore[method-assign]
ApiAnita.extraerMensajeError = staticmethod(ApiAnita.extraer_mensaje_error)  # type: ignore[method-assign]
ApiAnita.decodificarListaFilas = staticmethod(ApiAnita.decodificar_lista_filas)  # type: ignore[method-assign]
ApiAnita.primeraFilaLista = staticmethod(ApiAnita.primera_fila_lista)  # type: ignore[method-assign]


if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print("Uso: python api_anita.py <json_payload>")
        print('Ej.: python api_anita.py \'{"acc":"list","tabla":"x","campos":"id"}\'')
        sys.exit(1)

    api = ApiAnita()
    payload = json.loads(sys.argv[1])
    print(api.api_call(payload))

File: app/ApiAnita.php
<?php

namespace App;

use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Log;

class ApiAnita {
    protected string $fecha;

    protected string $servidorAnita;

    public function __construct()
    {
        $this->fecha = date('YmdHisu').'_'.random_int(0, 9999);
        $this->servidorAnita = (string) config('anita.ip', '');
    }

    /**
     * Resuelve host del bridge desde clave .env (ej. LOCAL_IP) o valor directo host:puerto.
     */
    public static function resolverHost(?string $claveServidor = null): string
    {
        if ($claveServidor !== null && trim($claveServidor) !== '') {
            $clave = strtoupper(trim($claveServidor));
            $mapa = config('anita.servidores', []);
            if (isset($mapa[$clave]) && trim((string) $mapa[$clave]) !== '') {
                return trim((string) $mapa[$clave]);
            }
            if (preg_match('#^[\w.\-]+(:\d+)?$#', trim($claveServidor))) {
                return trim($claveServidor);
            }
        }

        return trim((string) config('anita.ip', ''));
    }

    /**
     * URL del bridge HTTP (apiERP.php).
     */
    public static function urlBridge(?string $host = null): string
    {
        $host = trim((string) ($host ?? self::resolverHost()));
        if ($host === '') {
            throw new \RuntimeException(
                'ANITA_IP no está configurado. Defínalo en .env y ejecute php artisan config:cache (o config:clear).'
            );
        }

        $script = ltrim((string) config('anita.api_script', 'apiERP.php'), '/');

        if (preg_match('#^https?://#i', $host)) {
            $base = rtrim($host, '/');
            if (preg_match('#\.php(\?|$)#i', $base)) {
                return $base;
            }

            return $base.'/'.$script;
        }

        return 'http://'.$host.'/'.$script;
    }

    private static function resolverIfxServer(?string $claveIfx = null): string
    {
        if ($claveIfx !== null && trim($claveIfx) !== '') {
            $clave = strtoupper(trim($claveIfx));
            $mapa = config('anita.ifx_servers', []);
            if (isset($mapa[$clave]) && trim((string) $mapa[$clave]) !== '') {
                return trim((string) $mapa[$clave]);
            }
            if (preg_match('#^[\w.\-]+$#', trim($claveIfx))) {
                return trim($claveIfx);
            }
        }

        return trim((string) config('anita.ifx_server', ''));
    }

    public function apiCallHttp($data){
        $acc = (string) ($data['acc'] ?? '');

        if (isset($data['servidor'])) {
            $this->servidorAnita = self::resolverHost((string) $data['servidor']);
        }

        if (isset($data['ifx_server'])) {
            $data['IFX_SERVER'] = self::resolverIfxServer((string) $data['ifx_server']);
        } else {
            $data['IFX_SERVER'] = self::resolverIfxServer();
        }

        $bdd = (string) config('anita.bdd', 'ventas');
        $bddPath = rtrim((string) ($data['path_sistema'] ?? config('anita.bdd_path', '')), '/');
        $sistema = (isset($data['sistema']) && trim((string) $data['sistema']) !== '')
            ? trim((string) $data['sistema'])
            : $bdd;
        $data['sistema'] = $sistema;
        $data['DB_NAME'] = $sistema;
        $data['IFX_DB_PATH'] = ($bddPath !== '' ? $bddPath.'/' : '').$sistema;
        if ($bddPath !== '') {
            $data['path_sistema'] = $bddPath;
        }

        try {
            $url = self::urlBridge($this->servidorAnita);
        } catch (\RuntimeException $e) {
            return json_encode(['Error' => 'Bridge HTTP Anita: '.$e->getMessage()]);
        }

        $curl = curl_init();
        $payload = json_encode($data);

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url, CURLOPT_RETURNTRANSFER => true, CURLOPT_HEADER => false, CURLOPT_CUSTOMREQUEST => "POST", CURLOPT_POSTFIELDS => $payload
        ));
        curl_setopt($curl, CURLOPT_HTTPHEADER, array( 'Accept: application/json', 'Content-Type: application/json' )   );
        $response = curl_exec($curl);
        if (curl_errno($curl)) {
            $errorMsg = curl_error($curl);
            curl_close($curl);

            return json_encode(['Error' => 'Bridge HTTP Anita: '.$errorMsg]);
        }
        curl_close($curl);

        $trimResponse = trim((string) $response);
        if ($response === false || $trimResponse === '') {
            // acc=list sin filas: el bridge puede responder vacío; no es error.
            if ($acc === 'list') {
                return '[]';
            }
            // insert/update/delete exitosos en legacy devuelven []; el HTTP a veces body vacío.
            if (in_array($acc, ['insert', 'update', 'delete'], true)) {
                return '[]';
            }

            return json_encode(['Error' => 'Bridge HTTP Anita: respuesta vacía']);
        }

        // Quita warnings HTML del bridge y devuelve JSON limpio en consultas.
        if (in_array($acc, ['list', 'customSql'], true)) {
            return json_encode(self::decodificarListaFilas($trimResponse));
        }

        return $response;
    }

    /**
     * Quita advertencias PHP del body del bridge (insert/update) sin ocultar errores Informix.
     */
    public static function limpiarRespuestaBridgeEscritura(string $respuesta): string
    {
        $lineas = preg_split('/\R/', $respuesta) ?: [];
        $limpias = [];
        foreach ($lineas as $linea) {
            if (preg_match('/^\s*(?:Warning|Notice|Deprecated|Fatal error)\b/i', $linea)) {
                continue;
            }
            $limpias[] = $linea;
        }

        return trim(implode("\n", $limpias));
    }

    /**
     * Respuesta legacy del bridge en INSERT/UPDATE/DELETE exitoso.
     */
    public static function respuestaBridgeEscrituraExitosa(string $respuesta): bool
    {
        $limpia = self::limpiarRespuestaBridgeEscritura($respuesta);

        return $limpia !== ''
            && preg_match('/\d+\s+row\(s\)\s+(?:inserted|updated|deleted)\b/i', $limpia) === 1;
    }

    /**
     * Detecta error en la respuesta del bridge (HTTP o legacy).
     * [] es válido: lista sin filas (consulta) o OK en insert/update (bridge legacy).
     */
    public static function extraerMensajeError(?string $respuesta): ?string
    {
        if ($respuesta === null) {
            return 'Sin respuesta del bridge Anita';
        }

        $trim = trim($respuesta);
        if ($trim === '') {
            return 'Respuesta vacía del bridge Anita';
        }

        if ($trim === '[]' || $trim === '{}') {
            return null;
        }

        $decoded = json_decode($trim, true);
        if (is_array($decoded)) {
            if (isset($decoded['Error']) && (string) $decoded['Error'] !== '') {
                return (string) $decoded['Error'];
            }
            if (isset($decoded['error']) && (string) $decoded['error'] !== '') {
                return (string) $decoded['error'];
            }

            return null;
        }

        $limpia = self::limpiarRespuestaBridgeEscritura($trim);
        if (self::respuestaBridgeEscrituraExitosa($trim)) {
            return null;
        }

        if (strcasecmp($limpia, 'Error') === 0) {
            return 'Error en ejecución SQL Informix (revise el archivo .ret en el servidor Anita)';
        }

        if (stripos($trim, '<b>warning</b>') !== false || stripos($trim, '<b>fatal error</b>') !== false) {
            return strip_tags(html_entity_decode($trim));
        }

        if ($limpia === '' && preg_match('/\b(?:Warning|Notice|Fatal error)\b/i', $trim)) {
            return 'Advertencia PHP en bridge Anita (actualice apiERP.php en el servidor)';
        }

        if (stripos($limpia, 'error') !== false) {
            return $limpia;
        }

        return null;
    }

    /**
     * Ejecuta escritura en Informix vía bridge y falla si la respuesta indica error.
     */
    public function apiCallEscritura(array $payload, ?string $contexto = null, string $logEvento = 'anita_bridge.fallo'): string
    {
        if ($contexto === null || $contexto === '') {
            $tabla = $payload['tabla'] ?? 'sql';
            $acc = $payload['acc'] ?? '';
            $contexto = trim($tabla.' '.$acc);
        }

        $raw = (string) $this->apiCall($payload);
        $err = self::extraerMensajeError($raw === '' ? null : $raw);
        if ($err !== null) {
            Log::warning($logEvento, [
                'contexto' => $contexto,
                'tabla' => $payload['tabla'] ?? null,
                'acc' => $payload['acc'] ?? null,
                'mensaje' => $err,
            ]);
            throw new \RuntimeException('Error al grabar en Anita ('.$contexto.'): '.$err);
        }

        return $raw;
    }

    /**
     * Normaliza la respuesta JSON de acc=list del bridge a lista de filas (objetos).
     * El bridge HTTP puede devolver un array o un único objeto por fila; el acceso con [0]
     * sobre stdClass provoca "Cannot use object of type stdClass as array".
     *
     * @return list<object>
     */
    /**
     * @return array{filas: list<object>, error_lectura: ?string}
     */
    public static function parsearRespuestaLista(?string $respuesta): array
    {
        if ($respuesta === null) {
            return ['filas' => [], 'error_lectura' => 'Sin respuesta del bridge Anita'];
        }

        $trim = trim($respuesta);
        if ($trim === '' || $trim === '[]') {
            return ['filas' => [], 'error_lectura' => null];
        }

        $err = self::extraerMensajeError($trim);
        if ($err !== null) {
            return ['filas' => [], 'error_lectura' => $err];
        }

        if (preg_match('/(\[[\s\S]*\])\s*$/', $trim, $m)) {
            $trim = $m[1];
        }

        $decoded = json_decode($trim, false);
        if ($decoded === null) {
            return ['filas' => [], 'error_lectura' => 'Respuesta Anita no parseable como JSON'];
        }

        if (is_array($decoded)) {
            return ['filas' => array_values($decoded), 'error_lectura' => null];
        }

        if (is_object($decoded)) {
            if (isset($decoded->Error) || isset($decoded->error)) {
                $msg = trim((string) ($decoded->Error ?? $decoded->error ?? 'Error Anita'));

                return ['filas' => [], 'error_lectura' => $msg !== '' ? $msg : 'Error Anita'];
            }

            return ['filas' => [$decoded], 'error_lectura' => null];
        }

        return ['filas' => [], 'error_lectura' => 'Formato de respuesta Anita no reconocido'];
    }

    public static function decodificarListaFilas(?string $respuesta): array
    {
        return self::parsearRespuestaLista($respuesta)['filas'];
    }

    /**
     * Primera fila de una respuesta acc=list, o null si no hay datos.
     */
    public static function primeraFilaLista(?string $respuesta): ?object
    {
        $filas = self::decodificarListaFilas($respuesta);

        return $filas === [] ? null : $filas[0];
    }

    private function usaBridgeHttp(): bool
    {
        $tipo = strtoupper(trim((string) config('anita.bridge_type', config('gastronomia.anita_bridge_type', 'HTTP'))));

        return $tipo === 'HTTP';
    }

    public function apiCall($data){
	//dd('anita.bridge'.env('ANITA_BRIDGE_TYPE'));
	//dd(env('DB_CONNECTION'));

        if ($this->usaBridgeHttp()) {
            return $this->apiCallHttp($data);
        }
        $puertoSsh = config('anita.puerto_ssh');
        $portSSH = ($puertoSsh == null ? '' : '-p '.$puertoSsh);
        $portSCP = ($puertoSsh == null ? '' : '-P '.$puertoSsh);
        $sql = $this->armarSql($data);
        $nomArch = $this->fecha.".sql";
        $logsDir = storage_path('logs');
        if (! is_dir($logsDir)) {
            File::ensureDirectoryExists($logsDir);
        }
        $pathArch = $logsDir.'/'.$nomArch;
        File::put($logsDir.'/'.$this->fecha.'.sql', $sql);
        
        shell_exec('scp '.$portSCP.' '.$pathArch.' sergio@'.$this->servidorAnita.':/home/sergio/tmp/'.$nomArch.' > /dev/null');
        shell_exec('ssh '.$portSSH.' sergio@'.$this->servidorAnita.' "cd /usr2/www/htdocs; ./apiERP.php '.config('anita.bdd').' /home/sergio/tmp/'.$nomArch.' '.$this->fecha.' > /dev/null"');
        shell_exec("ssh ".$portSSH." sergio@".$this->servidorAnita." \"rm /home/sergio/tmp/".$nomArch." > /dev/null\"");
        
        if($data['acc'] == "list" || $data['acc'] == "customSql"){
            $csvPath = $logsDir.'/'.$this->fecha.'.csv';
            $bddPathSsh = rtrim((string) config('anita.bdd_path', ''), '/');
            $bddSsh = (string) config('anita.bdd', 'ventas');
            shell_exec('scp '.$portSCP.' sergio@'.$this->servidorAnita.':'.$bddPathSsh.'/'.$bddSsh.'/'.$this->fecha.'.csv '.$csvPath.' > /dev/null 2>&1');
            shell_exec("ssh ".$portSSH." sergio@".$this->servidorAnita." \"cd ".$bddPathSsh.'/'.$bddSsh."; rm ".$this->fecha.".csv > /dev/null\"");

            if (! is_readable($csvPath)) {
                if (is_file($logsDir.'/'.$this->fecha.'.sql')) {
                    @unlink($logsDir.'/'.$this->fecha.'.sql');
                }

                return json_encode([]);
            }

            $dataArr = array();
            $archivo = fopen($csvPath, 'r');
            $camposArr = explode(",", $data['campos']);
            while ($linea = fgets($archivo)) {
                $registroAssoc = array();
                $lineaArr = (explode("|", $linea));
                foreach ($camposArr as $key => $value) {
                    $nombreAux = explode(" as ", $value);
                    if (count($nombreAux) == 2) {
                        $value = $nombreAux[1];
                    }else{
                        $nombreAux = explode(" AS ", $value); 
                        if (count($nombreAux) == 2) {
                            $value = $nombreAux[1];
                        }
                    }
                    $registroAssoc[trim($value)] = $lineaArr[$key];
                }
                $dataArr[] = $registroAssoc; 
            }
            fclose($archivo);

            @unlink($csvPath);
            @unlink($logsDir.'/'.$this->fecha.'.sql');
            //dd($dataArr);
            return json_encode($dataArr);
        }
        return json_encode(array());
    }
    
    public function armarSql($data){
        $data['where'] 		 = (array_key_exists('where', $data) ? $data['where'] : "");
        $data['campos'] 	 = (array_key_exists('campos', $data) ? $data['campos'] : "");
        $data['tabla'] 		 = (array_key_exists('tabla', $data) ? $data['tabla'] : "");
        $data['whereArmado'] = (array_key_exists('whereArmado', $data) ? $data['whereArmado'] : "");
        $data['orderBy'] 	 = (array_key_exists('orderBy', $data) ? " ORDER BY ".$data['orderBy'] : "");
        $data['groupBy'] 	 = (array_key_exists('groupBy', $data) ? " GROUP BY ".$data['groupBy'] : "");
        $data['valores'] 	 = (array_key_exists('valores', $data) ? $data['valores'] : "");

        switch ($data['acc']){
            case 'list':
                $first = '';
                if (! empty($data['limit'])) {
                    $first = trim((string) $data['limit']).' ';
                }
                $sql = "UNLOAD TO '".$this->fecha.".csv' DELIMITER '|' SELECT ".$first.$data['campos']." FROM ".$data['tabla']." ".$data['whereArmado']." ".$data['groupBy']." ".$data['orderBy'];
            break;
            case 'insert':
                $sql = "INSERT INTO ".$data['tabla']." (".$data['campos'].") VALUES (".$data['valores'].")";
            break;
            case 'update':
                $sql = "UPDATE ".$data['tabla']." SET ".$data['valores']." ".$data['whereArmado'];
            break;
            case 'delete':
                $sql = "DELETE FROM ".$data['tabla']." ".$data['whereArmado'];
            break;
            case 'customSql':
                $sql = "UNLOAD TO '".$this->fecha.".csv' DELIMITER '|' ".$data['customSql'];
            break;
        }
        $sql = trim(preg_replace('/\s\s+/', ' ', $sql));
        return $sql;
    }

    public function obtenerSiguienteNumerador($tabla, $campoId = "id"){
        $id = 1;
        $data = array( 'acc' => 'list', 'campos' => 'MAX('.$campoId.') AS id', 'tabla' => $tabla );
        $fila = self::primeraFilaLista($this->apiCall($data));
        if ($fila !== null && isset($fila->id) && (string) $fila->id !== '') {
            $id = (int) $fila->id + 1;
        }
        return $id;
    }
}
